### PR TITLE
GCP credentials file token source

### PR DESCRIPTION
**Pull request by curtisallen**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
Updates `client-go`'s GCP token source, allowing users to specify `credentials-file` when creating a token source. Currently, if you don't set "cmd-path" then client-go will use the Google Application Credentials defined by `google.FindDefaultCredentials`. This works well except if you want to allow a process to connect to multiple clusters there is no way to specify multiple Google Application Credentials sources. This change allows you to set `credentials-file` in the config for gcp auth to make it easier to support connecting to multiple clusters.

**Special notes for your reviewer**: https://github.com/kubernetes/kubernetes/commit/ed58f37e794053fb9911f154b50bd8e20feaa53c updates golang.org/x/oauth2 as there were some updates to the GCP provider. 
**Does this PR introduce a user-facing change?**:

```release-notes
NONE
```

### CODE REVIEW

**Comment by mikedanese** [comment]
Can you explain why setting `GOOGLE_APPLICATION_CREDENTIALS` to the path of the credentials file doesn't work for you?

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go [hunk 1]
```diff
@@ -153,6 +166,19 @@ func tokenSource(isCmd bool, gcpConfig map[string]string) (oauth2.TokenSource, e
 
 	// Google Application Credentials-based token source
 	scopes := parseScopes(gcpConfig)
+	// If "credentials-file" is set, load from file. Otherwise, use default
+	// token source.
+	if credentialsFile, ok := gcpConfig["credentials-file"]; ok {
+		data, err := ioutil.ReadFile(credentialsFile)
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: This is evaluated relative to the CWD of the kubectl which is strange. Should we disallow relative paths?

**Comment by curtisallen** [comment]
@mikedanese Thanks for taking a look! Setting `GOOGLE_APPLICATION_CREDENTIALS` works fine if you need to use the client targeting a single k8s cluster. I have a long-running process that needs to use `client-go` to interact with several k8s clusters so relying on an environment variable with credentials for a single k8s cluster wouldn't be feasible.

**Comment by mikedanese** [comment]
Note that we have utitilities for working with and actually make use of TokenSources internally in client-go already, e.g.:

https://github.com/kubernetes/client-go/blob/982e3f55001f505adeeafb7f559affe3c7594420/rest/config.go#L343

I'm definitely not opposed to this change. I'm just trying to understand how generally useful it is.

**Comment by mikedanese** [comment]
cc @liggitt

**Comment by mikedanese** [comment]
Why does your application need a different service identity per cluster?

**Comment by curtisallen** [comment]
> Why does your application need a different service identity per cluster?

Because the GKE clusters span GCP Projects so there are different service identities.

**Comment by curtisallen** [comment]
@mikedanese Thought I'd provide a little more context on this change. We'd like to use `client-go` to validate running pods across a fleet of GKE clusters spanning GCP projects. I could be wrong but if I was using the `exec`, or `azure` token sources this use case is supported.

**Comment by mikedanese** [comment]
Yes, I see that this is not supported using kubeconfig with the gcp auth provider. A single service account can be given permission to use clusters across many GCP projects however. I'm trying to understand why you need a service account per project.

**Comment by mikedanese** [comment]
Also, I'm just giving this a bit of soak time to see if anyone else has comments.

**Comment by curtisallen** [comment]
Thanks @mikedanese, AFAIK there’s no way to make a single GCP IAM Service account span GCP projects. If you happen to know of any documentation stateing otherwise I’d love to see it.

**Comment by mikedanese** [comment]
You can set policies on resources that reference service accounts outside the project of that resource. 
```
$ gcloud projects add-iam-policy-binding projectB \
    --member serviceAccount:mysvcacct@projA.gserviceaccount.com \
    --role roles/container.clusterAdmin
$ gcloud projects add-iam-policy-binding projectC \
    --member serviceAccount:mysvcacct@projA.gserviceaccount.com \
    --role roles/container.clusterAdmin
```

You could also use RBAC.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go [hunk 2]
```diff
@@ -140,6 +151,9 @@ func tokenSource(isCmd bool, gcpConfig map[string]string) (oauth2.TokenSource, e
 		if gcpConfig["scopes"] != "" {
 			return nil, fmt.Errorf("scopes can only be used when kubectl is using a gcp service account key")
 		}
+		if gcpConfig["credentials-file"] != "" {
```
Comment (jlowdermilk) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: It's prior art, but you probably want to pull this out of the `isCmd` block; callers won't set `cmd-path` and `credentials-file` at the same time.  While you're at it, you can simplify this by removing the `isCmd` parameter and the `isCmdTokenSource` helper, as it is redundant.

Comment (curtisallen) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: Thanks @jlowdermilk 
gotcha on removing the credentails-file check as for the other part of your comment are you suggesting I do something like this?
```golang
func tokenSource(gcpConfig map[string]string) (oauth2.TokenSource, error) {
 	// Command-based token source
 	if cmd := isCmdTokenSource(gcpConfig); len(cmd) != 0 {
                // check gcp cmd config here
        ...
```

Comment (jlowdermilk) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: Yes, but without the helper

```
func tokenSource(gcpConfig map[string]string) (oauth2.TokenSource, error) {
 	// Command-based token source
 	if cmd, ok := gcpConfig["cmd-path"]; ok {
                // check gcp cmd config here
        ...
```

`isCmdTokenSource` can be removed entirely, it's equivalent to map key lookup.

**Pull request closed without merging**